File: server/src/api/types/answer.js
const { 
  GraphQLObjectType, GraphQLInt, GraphQLString, 
  GraphQLFloat, GraphQLList
} = require('graphql');
const GraphQLQuestion = require('./question');
const GraphQLUser = require('./user');
const User = require('../../models/user');


module.exports = new GraphQLObjectType({
  name: 'Answer',
  description: 'This represents an Answer',
  fields() {
    const GraphQLComment = require('./comment'); // need for resolving cyclic dependency

    return {
      id: {
        type: GraphQLInt,
        resolve(answer) {
          return answer.id;
        }
      },
      user: {
        type: GraphQLUser,
        resolve(answer) {
          return answer.getUser();
        }
      },
      text: {
        type: GraphQLString,
        resolve(answer) {
          return answer.text;
        }
      },
      question: {
        type: GraphQLQuestion,
        resolve(answer) {
          return answer.getQuestion();
        }
      },
      comments: {
        type: new GraphQLList(GraphQLComment),
        resolve(answer) {
          return answer.getComments();
        }
      },
      likes: {
        type: new GraphQLList(GraphQLUser),
        async resolve(answer) {
          let likes = await answer.getLikes(),
            ids = likes.map(like => ({ id: like.user_id }));
          return User.findAll({ where: { $or: ids } });
        }
      },
      timestamp: {
        type: GraphQLFloat,
        resolve(answer) {
          return new Date(answer.created_at).getTime();
        }
      }
    };
  }
});

File: graphql_types/answer.go
package graphql_types

import (
	"fmt"
	"github.com/drvirtuozov/ask-a-question/db"
	"github.com/drvirtuozov/ask-a-question/models"
	"github.com/graphql-go/graphql"
)

var Answer = graphql.NewObject(graphql.ObjectConfig{
	Name:        "Answer",
	Description: "This represents an Answer",
	Fields: graphql.Fields{
		"id": &graphql.Field{
			Type: graphql.Int,
			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
				return p.Source.(*models.UserAnswer).ID, nil
			},
		},
		"text": &graphql.Field{
			Type: graphql.String,
			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
				return p.Source.(*models.UserAnswer).Text, nil
			},
		},
		"user": &graphql.Field{
			Type: User,
			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
				user := &models.User{}
				err := db.Conn.Model(p.Source).Related(user).Error

				if err != nil {
					return nil, err
				}

				return user, nil
			},
		},
		"question": &graphql.Field{
			Type: Question,
			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
				question := &models.UserQuestion{}
				err := db.Conn.Model(p.Source).Related(question).Error

				if err != nil {
					return nil, err
				}

				return question, nil
			},
		},
		"comments": &graphql.Field{
			Type: graphql.NewList(Comment),
			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
				comments := []*models.AnswerComment{}
				err := db.Conn.Model(p.Source).Related(&comments).Error

				if err != nil {
					return nil, err
				}

				return comments, nil
			},
		},
		"likes": &graphql.Field{
			Type: graphql.NewList(User),
			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
				likes := []*models.AnswerLike{}
				err := db.Conn.Model(p.Source).Related(&likes).Error

				if err != nil {
					return nil, err
				}

				if len(likes) != 0 {
					users := []*models.User{}
					query := fmt.Sprintf("id = %d", likes[0].UserId)

					for i := 1; i < len(likes); i++ {
						query += fmt.Sprintf(" OR id = %d", likes[i].UserId)
					}

					err = db.Conn.Find(&users, query).Error

					if err != nil {
						return nil, err
					}

					return users, nil
				}

				return nil, nil
			},
		},
		"timestamp": &graphql.Field{
			Type: graphql.String,
			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
				return p.Source.(*models.UserAnswer).CreatedAt, nil // change to unix timestamp later
			},
		},
	},
})
